src/Microsoft.Health.Api/Features/HealthChecks/CachedHealthCheck.cs

```diff
@@ -7,148 +7,139 @@
+        TimeSpan maxWait = canSkip ? TimeSpan.Zero : Timeout.InfiniteTimeSpan;
 
         try
         {
             // Note: Only the timespan controls whether or not the method returns true/false.
             //       If the token is canceled, the method throws instead an OperationCanceledException.
             if (!await _semaphore.WaitAsync(maxWait, cancellationToken).ConfigureAwait(false))
             {
-                return _cachedResult.Value;
+                return _cache.Result;
             }
         }
-        catch (OperationCanceledException oce)
+        catch (OperationCanceledException oce) when (!HasExpired(Clock.UtcNow))
         {
-            // Re-evaluate the current time and return the cached value if it's still valid
-            if (!IsCacheValid(Clock.UtcNow))
-            {
-                _logger.LogWarning(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}.");
-                throw;
-            }
-
-            return _cachedResult.Value;
+            _logger.LogWarning(oce, "Health check was canceled. Falling back to cache.");
+            return _cache.Result;
         }
 
         try
         {
             // Once we've entered the semaphore, check the cache again for its freshness
-            if (IsCacheFresh(currentTime))
+            if (IsUpToDate(Clock.UtcNow))
             {
-                return _cachedResult.Value;
+                return _cache.Result;
             }
 
-            using IServiceScope scope = _provider.CreateScope();
-
             HealthCheckResult result;
             try
             {
-                IHealthCheck check = _healthCheckFactory.Invoke(scope.ServiceProvider);
-                result = await check.CheckHealthAsync(context, cancellationToken).ConfigureAwait(false);
+                result = await _healthCheck.CheckHealthAsync(context, cancellationToken).ConfigureAwait(false);
             }
-            catch (OperationCanceledException oce)
+            catch (Exception ex) when (!HasExpired(Clock.UtcNow))
             {
-                // Re-evaluate the current time and return the cached value if it's still valid
-                if (!IsCacheValid(Clock.UtcNow))
+                if (ex is OperationCanceledException oce)
                 {
-                    _logger.LogWarning(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}.");
-                    throw;
+                    _logger.LogWarning(oce, "Health check was canceled. Falling back to cache.");
+                }
+                else
+                {
+                    _logger.LogError(ex, "Health check failed to complete. Falling back to cache.");
                 }
 
-                return _cachedResult.Value;
+                return _cache.Result;
             }
 
-            return UpdateCache(result).Value;
+            // Update cache based on the latest snapshot
+            return RefreshCache(result);
```

Comment:
It may make it more clear to have the `RefreshCache` just do it's updating the `_cache` object and just return `_cache.Result` like we do elsewhere?